integration-tests/src/tests/client/resharding.rs

```diff
@@ -1583,5 +1593,93 @@ fn test_latest_protocol_missing_chunks_high_missing_prob() {
     test_latest_protocol_missing_chunks(0.9, 27);
 }
 
+fn test_shard_layout_upgrade_error_handling_impl(
+    resharding_type: ReshardingType,
+    rng_seed: u64,
+    state_snapshot_enabled: bool,
+) {
+    init_test_logger();
+    tracing::info!(target: "test", "test_shard_layout_upgrade_simple_impl starting");
+
+    let genesis_protocol_version = get_genesis_protocol_version(&resharding_type);
+    let target_protocol_version = get_target_protocol_version(&resharding_type);
+
+    // setup
+    let epoch_length = 5;
+    let mut test_env = TestReshardingEnv::new(
+        epoch_length,
+        2,
+        2,
+        100,
+        None,
+        genesis_protocol_version,
+        rng_seed,
+        state_snapshot_enabled,
+        Some(resharding_type),
+    );
+    test_env.set_init_tx(vec![]);
+
+    let mut nonce = 100;
+    let genesis_hash = *test_env.env.clients[0].chain.genesis_block().hash();
+    let mut all_accounts: HashSet<_> = test_env.initial_accounts.clone().into_iter().collect();
+    let mut accounts_to_check: Vec<_> = vec![];
+    let initial_accounts = test_env.initial_accounts.clone();
+
+    // add transactions until after sharding upgrade finishes
+    for height in 2..3 * epoch_length {
+        let txs = generate_create_accounts_txs(
+            &mut test_env.rng,
+            genesis_hash,
+            &initial_accounts,
+            &mut accounts_to_check,
+            &mut all_accounts,
+            &mut nonce,
+            10,
+            true,
+        );
+
+        test_env.set_tx_at_height(height, txs);
+    }
+
+    let drop_chunk_condition = DropChunkCondition::new();
+    for _ in 1..4 * epoch_length {
+        let result = test_env.step_err(&drop_chunk_condition, target_protocol_version);
+        if let Err(err) = result {
+            tracing::info!(target: "test", ?err, "step failed, as expected");
+            return;
+        }
+        test_env.check_receipt_id_to_shard_id();
+        test_env.check_snapshot(state_snapshot_enabled);
+
+        // corrupt the state snapshot if available to make resharding fail
+        let client = test_env.env.clients.get(0).unwrap();
+        let runtime = &client.runtime_adapter;
+        let tries = runtime.get_tries();
+        let snapshot_hash = tries.get_state_snapshot_hash();
+        let Ok(snapshot_hash) = snapshot_hash else { continue };
```

Comment:
Actually, we might want to be super specific about when we initiate this. Should call this only for a specific height value instead of always checking if snapshot exits.

Reply:
I feel there isn't much downside to this impl and it is a bit more robust to changes in the test iteself. Any particular reason why you think it's worth being specific here?

Reply:
Yeah, it's fine, can leave as is... see if you'd like to integrate in the nit tho